File: backend/app/models/schemas/signal.py
from __future__ import annotations

from datetime import datetime, date
from typing import Any, Optional
from pydantic import BaseModel


class SignalListItem(BaseModel):
    id: str
    generated_at: datetime
    action: str
    instrument: Optional[str] = None
    strike_a: Optional[float] = None
    strike_b: Optional[float] = None
    expiry: Optional[date] = None
    quantity: Optional[int] = None
    composite_score: Optional[float] = None
    confidence: Optional[str] = None
    market_regime: Optional[str] = None

    class Config:
        from_attributes = True


class SignalResponse(BaseModel):
    id: str
    generated_at: datetime
    action: str
    instrument: Optional[str] = None
    strike_a: Optional[float] = None
    strike_b: Optional[float] = None
    expiry: Optional[date] = None
    quantity: Optional[int] = None
    premium_collected: Optional[float] = None
    max_profit: Optional[float] = None
    max_loss: Optional[float] = None
    breakeven_a: Optional[float] = None
    breakeven_b: Optional[float] = None
    probability_of_profit: Optional[float] = None
    expected_value: Optional[float] = None
    kelly_fraction: Optional[float] = None
    composite_score: Optional[float] = None
    confidence: Optional[str] = None
    iv_rank: Optional[float] = None
    iv_percentile: Optional[float] = None
    market_regime: Optional[str] = None
    market_bias: Optional[str] = None
    reasoning_plain: Optional[str] = None
    reasoning_detail: Optional[str] = None
    greeks: Optional[Any] = None
    score_breakdown: Optional[Any] = None
    alt_signal: Optional[Any] = None
    snapshot_id: Optional[int] = None
    pushed_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class SignalListResponse(BaseModel):
    items: list[SignalListItem]
    total: int
    page: int
    page_size: int

File: optigold_mobile/lib/core/api/models/signal.dart
class SignalModel {
  final String id;
  final DateTime generatedAt;
  final String action;
  final String? instrument;
  final double? strikeA;
  final double? strikeB;
  final String? expiry;
  final int? quantity;
  final double? premiumCollected;
  final double? maxProfit;
  final double? maxLoss;
  final double? breakevenA;
  final double? breakevenB;
  final double? probabilityOfProfit;
  final double? expectedValue;
  final double? kellyFraction;
  final double? compositeScore;
  final String? confidence;
  final double? ivRank;
  final double? ivPercentile;
  final String? marketRegime;
  final String? marketBias;
  final String? reasoningPlain;
  final String? reasoningDetail;
  final Map<String, dynamic>? greeks;
  final Map<String, dynamic>? scoreBreakdown;
  final Map<String, dynamic>? altSignal;

  const SignalModel({
    required this.id,
    required this.generatedAt,
    required this.action,
    this.instrument,
    this.strikeA,
    this.strikeB,
    this.expiry,
    this.quantity,
    this.premiumCollected,
    this.maxProfit,
    this.maxLoss,
    this.breakevenA,
    this.breakevenB,
    this.probabilityOfProfit,
    this.expectedValue,
    this.kellyFraction,
    this.compositeScore,
    this.confidence,
    this.ivRank,
    this.ivPercentile,
    this.marketRegime,
    this.marketBias,
    this.reasoningPlain,
    this.reasoningDetail,
    this.greeks,
    this.scoreBreakdown,
    this.altSignal,
  });

  factory SignalModel.fromJson(Map<String, dynamic> j) => SignalModel(
        id:                  j['id'] as String,
        generatedAt:         DateTime.parse(j['generated_at'] as String),
        action:              j['action'] as String,
        instrument:          j['instrument'] as String?,
        strikeA:             (j['strike_a'] as num?)?.toDouble(),
        strikeB:             (j['strike_b'] as num?)?.toDouble(),
        expiry:              j['expiry'] as String?,
        quantity:            j['quantity'] as int?,
        premiumCollected:    (j['premium_collected'] as num?)?.toDouble(),
        maxProfit:           (j['max_profit'] as num?)?.toDouble(),
        maxLoss:             (j['max_loss'] as num?)?.toDouble(),
        breakevenA:          (j['breakeven_a'] as num?)?.toDouble(),
        breakevenB:          (j['breakeven_b'] as num?)?.toDouble(),
        probabilityOfProfit: (j['probability_of_profit'] as num?)?.toDouble(),
        expectedValue:       (j['expected_value'] as num?)?.toDouble(),
        kellyFraction:       (j['kelly_fraction'] as num?)?.toDouble(),
        compositeScore:      (j['composite_score'] as num?)?.toDouble(),
        confidence:          j['confidence'] as String?,
        ivRank:              (j['iv_rank'] as num?)?.toDouble(),
        ivPercentile:        (j['iv_percentile'] as num?)?.toDouble(),
        marketRegime:        j['market_regime'] as String?,
        marketBias:          j['market_bias'] as String?,
        reasoningPlain:      j['reasoning_plain'] as String?,
        reasoningDetail:     j['reasoning_detail'] as String?,
        greeks:              j['greeks'] as Map<String, dynamic>?,
        scoreBreakdown:      j['score_breakdown'] as Map<String, dynamic>?,
        altSignal:           j['alt_signal'] as Map<String, dynamic>?,
      );

  // Derived helpers
  bool get isBullish   => marketBias?.toLowerCase() == 'bullish';
  bool get isBearish   => marketBias?.toLowerCase() == 'bearish';
  bool get isHighConf  => confidence?.toLowerCase() == 'high';
  bool get isMedConf   => confidence?.toLowerCase() == 'medium';

  String get formattedRegime {
    if (marketRegime == null) return 'Unknown';
    return marketRegime!.replaceAll('_', ' ').split(' ')
        .map((w) => w[0].toUpperCase() + w.substring(1))
        .join(' ');
  }

  double get pop => probabilityOfProfit ?? 0.0;
  double get score => compositeScore ?? 0.0;
}

class SignalListItem {
  final String id;
  final DateTime generatedAt;
  final String action;
  final String? instrument;
  final double? strikeA;
  final double? strikeB;
  final String? expiry;
  final int? quantity;
  final double? compositeScore;
  final String? confidence;
  final String? marketRegime;

  const SignalListItem({
    required this.id,
    required this.generatedAt,
    required this.action,
    this.instrument,
    this.strikeA,
    this.strikeB,
    this.expiry,
    this.quantity,
    this.compositeScore,
    this.confidence,
    this.marketRegime,
  });

  factory SignalListItem.fromJson(Map<String, dynamic> j) => SignalListItem(
        id:             j['id'] as String,
        generatedAt:    DateTime.parse(j['generated_at'] as String),
        action:         j['action'] as String,
        instrument:     j['instrument'] as String?,
        strikeA:        (j['strike_a'] as num?)?.toDouble(),
        strikeB:        (j['strike_b'] as num?)?.toDouble(),
        expiry:         j['expiry'] as String?,
        quantity:       j['quantity'] as int?,
        compositeScore: (j['composite_score'] as num?)?.toDouble(),
        confidence:     j['confidence'] as String?,
        marketRegime:   j['market_regime'] as String?,
      );
}
